README.md
# aurelia-html-import-template-loader

[![npm Version](https://img.shields.io/npm/v/aurelia-html-import-template-loader.svg)](https://www.npmjs.com/package/aurelia-html-import-template-loader)
[![ZenHub](https://raw.githubusercontent.com/ZenHubIO/support/master/zenhub-badge.png)](https://zenhub.io)
[![Join the chat at https://gitter.im/aurelia/discuss](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/aurelia/discuss?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![CircleCI](https://circleci.com/gh/aurelia/html-import-template-loader.svg?style=shield)](https://circleci.com/gh/aurelia/html-import-template-loader)

This library is part of the [Aurelia](http://www.aurelia.io/) platform and contains an implementation of the TemplateLoader interface for HTMLImport loading which works with system.js and all require-based loader APIs.

> To keep up to date on [Aurelia](http://www.aurelia.io/), please visit and subscribe to [the official blog](http://blog.durandal.io/) and [our email list](http://durandal.us10.list-manage1.com/subscribe?u=dae7661a3872ee02b519f6f29&id=3de6801ccc). We also invite you to [follow us on twitter](https://twitter.com/aureliaeffect). If you have questions, please [join our community on Gitter](https://gitter.im/aurelia/discuss). If you would like to have deeper insight into our development process, please install the [ZenHub](https://zenhub.io) Chrome or Firefox Extension and visit any of our repository's boards. You can get an overview of all Aurelia work by visiting [the framework board](https://github.com/aurelia/framework#boards).

## Platform Support

This library can be used in the **browser** only.

## Building The Code

To build the code, follow these steps.

1. Ensure that [NodeJS](http://nodejs.org/) is installed. This provides the platform on which the build tooling runs.
2. From the project folder, execute the following command:

  ```shell
  npm install
  ```
3. Ensure that [Gulp](http://gulpjs.com/) is installed. If you need to install it, use the following command:

  ```shell
  npm install -g gulp
  ```
4. To build the code, you can now run:

  ```shell
  gulp build
  ```
5. You will find the compiled code in the `dist` folder, available in three module formats: AMD, CommonJS and ES6.

6. See `gulpfile.js` for other tasks related to generating the docs and linting.

## Running The Tests

To run the unit tests, first ensure that you have followed the steps above in order to install all dependencies and successfully build the library. Once you have done that, proceed with these additional steps:

1. Ensure that the [Karma](http://karma-runner.github.io/) CLI is installed. If you need to install it, use the following command:

  ```shell
  npm install -g karma-cli
  ```
2. Ensure that [jspm](http://jspm.io/) is installed. If you need to install it, use the following commnand:

  ```shell
  npm install -g jspm
  ```
3. Install the client-side dependencies with jspm:

  ```shell
  jspm install
  ```

4. You can now run the tests with this command:

  ```shell
  karma start
  ```


doc/CHANGELOG.md
<a name="1.0.0"></a>
# [1.0.0](https://github.com/aurelia/html-import-template-loader/compare/1.0.0-rc.1.0.0...v1.0.0) (2016-07-27)



<a name="1.0.0-rc.1.0.0"></a>
# [1.0.0-rc.1.0.0](https://github.com/aurelia/html-import-template-loader/compare/1.0.0-beta.2.0.1...v1.0.0-rc.1.0.0) (2016-06-22)



### 1.0.0-beta.1.2.2 (2016-05-17)


#### Bug Fixes

* **all:** update more to use PAL ([1eb19a2d](http://github.com/aurelia/html-import-template-loader/commit/1eb19a2da8ad55c59de295eb0dd26b0a0518a4e3))
* Address various module/package name issues.

### 1.0.0-beta.1.2.1 (2016-05-10)


### 1.0.0-beta.1.2.0 (2016-03-22)

* Update to Babel 6

### 1.0.0-beta.1.1.3 (2016-02-08)


#### Bug Fixes

* **loader:** missing context variable ([8aa788ae](http://github.com/aurelia/html-import-template-loader/commit/8aa788aef594d3d8eb0aa3d6d02c69e795ae8301), closes [#7](http://github.com/aurelia/html-import-template-loader/issues/7))


### 1.0.0-beta.1.1.0 (2016-01-29)


#### Features

* **all:** update jspm meta; core-js; aurelia deps ([b4faa50e](http://github.com/aurelia/html-import-template-loader/commit/b4faa50ee4457710e6d488d66fba373479b4b96b))


### 1.0.0-beta.1.0.1 (2016-01-08)


#### Features

* **all:** switch to normalize api from normalizeSync ([aa8870d9](http://github.com/aurelia/html-import-template-loader/commit/aa8870d93c05e730842bd87b3a555c009aaca34e))


### 1.0.0-beta.1 (2015-11-16)


## 0.4.0 (2015-11-10)


#### Bug Fixes

* **all:** update to latest template entry api ([50234217](http://github.com/aurelia/html-import-template-loader/commit/5023421742a44e856a579d0c46d590358554dec6))


## 0.3.0 (2015-10-13)


#### Bug Fixes

* **all:**
  * incorrect module id to normalize polyfill with ([fb0a6b33](http://github.com/aurelia/html-import-template-loader/commit/fb0a6b33d55fe9f068d2bb8af5c897ce73f24145))
  * correct template id normalization for bundled templates ([181fcc5d](http://github.com/aurelia/html-import-template-loader/commit/181fcc5dbb580086844c01c0de93ba0644ad4eef))
* **package:** incorrect main ([3a68e742](http://github.com/aurelia/html-import-template-loader/commit/3a68e742225ce186dc7229e7534d34c6ed5729e8))


#### Features

* **all:**
  * update to pal ([e397cd1f](http://github.com/aurelia/html-import-template-loader/commit/e397cd1fa799b53a64708a4501eb2081b16aea08))
  * use new loader apis to eliminate global System calls ([54b5dce2](http://github.com/aurelia/html-import-template-loader/commit/54b5dce2d669b20bd36ff337a73afb7534c38606))
  * added and configured the html import template loader ([d6923248](http://github.com/aurelia/html-import-template-loader/commit/d6923248b1311a41ffc2ea771b4e06ceb01d6b20))


### 0.2.2 (2015-09-22)


#### Bug Fixes

* **all:** incorrect module id to normalize polyfill with ([fb0a6b33](http://github.com/aurelia/html-import-template-loader/commit/fb0a6b33d55fe9f068d2bb8af5c897ce73f24145))


### 0.2.1 (2015-09-09)


#### Bug Fixes

* **all:** correct template id normalization for bundled templates ([181fcc5d](http://github.com/aurelia/html-import-template-loader/commit/181fcc5dbb580086844c01c0de93ba0644ad4eef))
* **package:** incorrect main ([3a68e742](http://github.com/aurelia/html-import-template-loader/commit/3a68e742225ce186dc7229e7534d34c6ed5729e8))


## 0.2.0 (2015-09-05)


#### Features

* **all:** added and configured the html import template loader ([d6923248](http://github.com/aurelia/html-import-template-loader/commit/d6923248b1311a41ffc2ea771b4e06ceb01d6b20))


config.js
System.config({
  defaultJSExtensions: true,
  paths: {
    "github:*": "jspm_packages/github/*",
    "npm:*": "jspm_packages/npm/*"
  },

  map: {
    "aurelia-loader": "npm:aurelia-loader@1.0.0",
    "aurelia-pal": "npm:aurelia-pal@1.0.0",
    "babel": "npm:babel-core@5.8.38",
    "babel-runtime": "npm:babel-runtime@5.8.38",
    "core-js": "npm:core-js@2.4.1",
    "webcomponentsjs": "npm:webcomponents.js@0.7.22",
    "github:jspm/nodelibs-assert@0.1.0": {
      "assert": "npm:assert@1.4.1"
    },
    "github:jspm/nodelibs-buffer@0.1.0": {
      "buffer": "npm:buffer@3.6.0"
    },
    "github:jspm/nodelibs-events@0.1.1": {
      "events": "npm:events@1.0.2"
    },
    "github:jspm/nodelibs-os@0.1.0": {
      "os-browserify": "npm:os-browserify@0.1.2"
    },
    "github:jspm/nodelibs-path@0.1.0": {
      "path-browserify": "npm:path-browserify@0.0.0"
    },
    "github:jspm/nodelibs-process@0.1.2": {
      "process": "npm:process@0.11.6"
    },
    "github:jspm/nodelibs-stream@0.1.0": {
      "stream-browserify": "npm:stream-browserify@1.0.0"
    },
    "github:jspm/nodelibs-string_decoder@0.1.0": {
      "string_decoder": "npm:string_decoder@0.10.31"
    },
    "github:jspm/nodelibs-tty@0.1.0": {
      "tty-browserify": "npm:tty-browserify@0.0.0"
    },
    "github:jspm/nodelibs-util@0.1.0": {
      "util": "npm:util@0.10.3"
    },
    "github:jspm/nodelibs-vm@0.1.0": {
      "vm-browserify": "npm:vm-browserify@0.0.4"
    },
    "npm:assert@1.4.1": {
      "assert": "github:jspm/nodelibs-assert@0.1.0",
      "buffer": "github:jspm/nodelibs-buffer@0.1.0",
      "process": "github:jspm/nodelibs-process@0.1.2",
      "util": "npm:util@0.10.3"
    },
    "npm:aurelia-loader@1.0.0": {
      "aurelia-metadata": "npm:aurelia-metadata@1.0.0",
      "aurelia-path": "npm:aurelia-path@1.0.0"
    },
    "npm:aurelia-metadata@1.0.0": {
      "aurelia-pal": "npm:aurelia-pal@1.0.0"
    },
    "npm:babel-runtime@5.8.38": {
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:buffer@3.6.0": {
      "base64-js": "npm:base64-js@0.0.8",
      "child_process": "github:jspm/nodelibs-child_process@0.1.0",
      "fs": "github:jspm/nodelibs-fs@0.1.2",
      "ieee754": "npm:ieee754@1.1.6",
      "isarray": "npm:isarray@1.0.0",
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:core-js@2.4.1": {
      "fs": "github:jspm/nodelibs-fs@0.1.2",
      "path": "github:jspm/nodelibs-path@0.1.0",
      "process": "github:jspm/nodelibs-process@0.1.2",
      "systemjs-json": "github:systemjs/plugin-json@0.1.2"
    },
    "npm:core-util-is@1.0.2": {
      "buffer": "github:jspm/nodelibs-buffer@0.1.0"
    },
    "npm:inherits@2.0.1": {
      "util": "github:jspm/nodelibs-util@0.1.0"
    },
    "npm:os-browserify@0.1.2": {
      "os": "github:jspm/nodelibs-os@0.1.0"
    },
    "npm:path-browserify@0.0.0": {
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:process@0.11.6": {
      "assert": "github:jspm/nodelibs-assert@0.1.0",
      "fs": "github:jspm/nodelibs-fs@0.1.2",
      "vm": "github:jspm/nodelibs-vm@0.1.0"
    },
    "npm:readable-stream@1.1.14": {
      "buffer": "github:jspm/nodelibs-buffer@0.1.0",
      "core-util-is": "npm:core-util-is@1.0.2",
      "events": "github:jspm/nodelibs-events@0.1.1",
      "inherits": "npm:inherits@2.0.1",
      "isarray": "npm:isarray@0.0.1",
      "process": "github:jspm/nodelibs-process@0.1.2",
      "stream-browserify": "npm:stream-browserify@1.0.0",
      "string_decoder": "npm:string_decoder@0.10.31"
    },
    "npm:stream-browserify@1.0.0": {
      "events": "github:jspm/nodelibs-events@0.1.1",
      "inherits": "npm:inherits@2.0.1",
      "readable-stream": "npm:readable-stream@1.1.14"
    },
    "npm:string_decoder@0.10.31": {
      "buffer": "github:jspm/nodelibs-buffer@0.1.0"
    },
    "npm:util@0.10.3": {
      "inherits": "npm:inherits@2.0.1",
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:vm-browserify@0.0.4": {
      "indexof": "npm:indexof@0.0.1"
    },
    "npm:webcomponents.js@0.7.22": {
      "buffer": "github:jspm/nodelibs-buffer@0.1.0",
      "child_process": "github:jspm/nodelibs-child_process@0.1.0",
      "events": "github:jspm/nodelibs-events@0.1.1",
      "fs": "github:jspm/nodelibs-fs@0.1.2",
      "os": "github:jspm/nodelibs-os@0.1.0",
      "path": "github:jspm/nodelibs-path@0.1.0",
      "process": "github:jspm/nodelibs-process@0.1.2",
      "stream": "github:jspm/nodelibs-stream@0.1.0",
      "string_decoder": "github:jspm/nodelibs-string_decoder@0.1.0",
      "systemjs-json": "github:systemjs/plugin-json@0.1.2",
      "tty": "github:jspm/nodelibs-tty@0.1.0",
      "util": "github:jspm/nodelibs-util@0.1.0"
    }
  }
});


src/html-import-template-loader.js
import {TemplateRegistryEntry, Loader} from 'aurelia-loader';
import {FEATURE} from 'aurelia-pal';

/**
* An implementation of the TemplateLoader interface implemented using HTML Imports.
* Suitable for use when integrating with Polymer.
*/
export class HTMLImportTemplateLoader {
  /**
   * Creates an instance of HTMLImportTemplateLoader.
   */
  constructor() {
    this.needsBundleCheck = true;
    this.onBundleReady = null;
  }

  /**
  * Loads a template.
  * @param loader The loader that is requesting the template load.
  * @param entry The TemplateRegistryEntry to load and populate with a template.
  * @return A promise which resolves when the TemplateRegistryEntry is loaded with a template.
  */
  loadTemplate(loader: Loader, entry: TemplateRegistryEntry): Promise<any> {
    return this._tryFindTemplateInBundle(loader, entry).then(found => {
      return found ? entry : this._importDocument(entry).then(doc => this._findTemplate(doc, entry));
    });
  }

  _tryFindTemplateInBundle(loader, entry) {
    if (this.bundle) {
      return this._tryGetTemplateFromBundle(entry);
    } else if (this.onBundleReady) {
      return this.onBundleReady.then(() => this._tryGetTemplateFromBundle(entry));
    } else if (this.needsBundleCheck) {
      return this._loadBundle(loader, entry);
    }

    return Promise.resolve(false);
  }

  _loadBundle(loader, entry) {
    let bundleLink = document.querySelector('link[aurelia-view-bundle]');
    this.needsBundleCheck = false;

    if (bundleLink) {
      this.onBundleReady = this._importBundle(bundleLink).then(doc => {
        return this._normalizeTemplateIds(loader, doc).then(() => {
          this.bundle = doc;
          this.onBundleReady = null;
        });
      });

      return this.onBundleReady.then(() => this._tryGetTemplateFromBundle(entry));
    }

    return Promise.resolve(false);
  }

  _importDocument(entry) {
    return new Promise((resolve, reject) => {
      let frag = document.createDocumentFragment();
      let link = document.createElement('link');

      link.rel = 'import';
      link.href = entry.address;
      frag.appendChild(link);

      this._importElements(frag, link, () => resolve(link.import));
    });
  }

  _findTemplate(doc, entry) {
    let template = doc.getElementsByTagName('template')[0];

    if (!template) {
      throw new Error(`There was no template element found in '${entry.address}'.`);
    }

    entry.template = FEATURE.ensureHTMLTemplateElement(template);
  }

  _tryGetTemplateFromBundle(entry) {
    let found = this.bundle.getElementById(entry.address);

    if (found) {
      entry.template = FEATURE.ensureHTMLTemplateElement(found);
      return Promise.resolve(true);
    }

    return Promise.resolve(false);
  }

  _importBundle(link) {
    return new Promise((resolve, reject) => {
      if (link.import) {
        resolve(link.import);
      } else {
        this._importElements(null, link, () => resolve(link.import));
      }
    });
  }

  _normalizeTemplateIds(loader, doc) {
    let templates = doc.getElementsByTagName('template');
    let i = templates.length;
    let all = [];

    while (i--) {
      let current = templates[i];
      let id = current.getAttribute('id');

      if (id !== null && id !== undefined) {
        all.push(normalizeTemplateId(loader, id, current));
      }
    }

    return Promise.all(all);
  }

  _importElements(frag, link, callback) {
    if (frag) {
      document.head.appendChild(frag);
    }

    if (window.Polymer && Polymer.whenReady ) {
      Polymer.whenReady(callback);
    } else {
      link.addEventListener('load', callback);
    }
  }
}

function normalizeTemplateId(loader, id, current) {
  let beforeNormalize = id + '!template-registry-entry';

  return loader.normalize(beforeNormalize).then(afterNormalize => {
    current.setAttribute('id', afterNormalize.replace('!template-registry-entry', ''));
  });
}

/**
 * Configuires the HTMLImportTemplateLoader as the default loader for Aurelia.
 * @param config The FrameworkConfiguration instance.
 */
export function configure(config: Object): Promise<void> {
  config.aurelia.loader.useTemplateLoader(new HTMLImportTemplateLoader());

  // Check for native or polyfilled HTML imports support.
  if (!('import' in document.createElement('link')) && !('HTMLImports' in window)) {
    return config.aurelia.loader.normalize('aurelia-html-import-template-loader').then(name => {
      return config.aurelia.loader.normalize('webcomponentsjs/HTMLImports.min', name);
    }).then(importsName => config.aurelia.loader.loadModule(importsName));
  }
}


src/aurelia-html-import-template-loader.js
import {TemplateRegistryEntry, Loader} from 'aurelia-loader';
import {FEATURE, DOM, PLATFORM} from 'aurelia-pal';

/**
* An implementation of the TemplateLoader interface implemented using HTML Imports.
* Suitable for use when integrating with Polymer.
*/
export class HTMLImportTemplateLoader {

  /**
   * Creates an instance of HTMLImportTemplateLoader.
   */
  constructor(linkHrefPrefix) {
    this.linkHrefPrefix = linkHrefPrefix || '';
    this.needsBundleCheck = true;
    this.onBundleReady = null;
  }

  /**
  * Loads a template.
  * @param loader The loader that is requesting the template load.
  * @param entry The TemplateRegistryEntry to load and populate with a template.
  * @return A promise which resolves when the TemplateRegistryEntry is loaded with a template.
  */
  loadTemplate(loader: Loader, entry: TemplateRegistryEntry): Promise<any> {
    return this._tryFindTemplateInBundle(loader, entry).then(found => {
      return found ? entry : this._importDocument(entry).then(doc => this._findTemplate(doc, entry));
    });
  }

  _tryFindTemplateInBundle(loader, entry) {
    if (this.bundle) {
      return this._tryGetTemplateFromBundle(entry);
    } else if (this.onBundleReady) {
      return this.onBundleReady.then(() => this._tryGetTemplateFromBundle(entry));
    } else if (this.needsBundleCheck) {
      return this._loadBundle(loader, entry);
    }

    return Promise.resolve(false);
  }

  _loadBundle(loader, entry) {
    let bundleLink = document.querySelector('link[aurelia-view-bundle]');
    this.needsBundleCheck = false;

    if (bundleLink) {
      this.onBundleReady = this._importBundle(bundleLink).then(doc => {
        return this._normalizeTemplateIds(loader, doc).then(() => {
          this.bundle = doc;
          this.onBundleReady = null;
        });
      });

      return this.onBundleReady.then(() => this._tryGetTemplateFromBundle(entry));
    }

    return Promise.resolve(false);
  }

  _importDocument(entry) {
    return new Promise((resolve, reject) => {
      let frag = DOM.createDocumentFragment();
      let link = DOM.createElement('link');

      link.rel = 'import';
      link.href = this.linkHrefPrefix + entry.address;
      frag.appendChild(link);

      this._importElements(frag, link, () => resolve(link.import));
    });
  }

  _findTemplate(doc, entry) {
    let template = doc.getElementsByTagName('template')[0];

    if (!template) {
      throw new Error(`There was no template element found in '${entry.address}'.`);
    }
    entry.template = FEATURE.ensureHTMLTemplateElement(template);
  }

  _tryGetTemplateFromBundle(entry) {
    let found = this.bundle.getElementById(entry.address);

    if (found) {
      entry.template = FEATURE.ensureHTMLTemplateElement(found);
      return Promise.resolve(true);
    }

    return Promise.resolve(false);
  }

  _importBundle(link) {
    return new Promise((resolve, reject) => {
      if (link.import) {
        resolve(link.import);
      } else {
        this._importElements(null, link, () => resolve(link.import));
      }
    });
  }

  _normalizeTemplateIds(loader, doc) {
    let templates = doc.getElementsByTagName('template');
    let i = templates.length;
    let all = [];

    while (i--) {
      let current = templates[i];
      let id = current.getAttribute('id');

      if (id !== null && id !== undefined) {
        all.push(normalizeTemplateId(loader, id, current));
      }
    }

    return Promise.all(all);
  }

  _importElements(frag, link, callback) {
    if (frag) {
      document.head.appendChild(frag);
    }

    if (PLATFORM.global.Polymer && Polymer.whenReady) {
      Polymer.whenReady(callback);
    } else {
      link.addEventListener('load', callback);
    }
  }
}

function normalizeTemplateId(loader, id, current) {
  let beforeNormalize = id + '!template-registry-entry';

  return loader.normalize(beforeNormalize).then(afterNormalize => {
    current.setAttribute('id', afterNormalize.replace('!template-registry-entry', ''));
  });
}

/**
 * Configuires the HTMLImportTemplateLoader as the default loader for Aurelia.
 * @param config The FrameworkConfiguration instance.
 */
export function configure(config: Object, inlineConfig: Object): Promise<void> {
  config.aurelia.loader.useTemplateLoader(new HTMLImportTemplateLoader(inlineConfig.linkHrefPrefix));

  // Check for native or polyfilled HTML imports support.
  if (!('import' in DOM.createElement('link')) && !('HTMLImports' in PLATFORM.global)) {
    return config.aurelia.loader.normalize('aurelia-html-import-template-loader').then(name => {
      return config.aurelia.loader.normalize('webcomponentsjs/HTMLImports.min', name);
    }).then(importsName => config.aurelia.loader.loadModule(importsName));
  }

  return undefined;
}


karma.conf.js
// Karma configuration
// Generated on Fri Dec 05 2014 16:49:29 GMT-0500 (EST)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jspm', 'jasmine'],

    jspm: {
      // Edit this to your needs
      loadFiles: ['test/**/*.spec.js'],
      serveFiles : ['src/**/*.js','test/fixtures/**/*.js']
    },


    // list of files / patterns to load in the browser
    files: [
      {pattern: 'test/fixtures/**/*.html', included: false}
    ],


    // list of files to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'test/**/*.js': ['babel'],
      'src/**/*.js': ['babel']
    },
    'babelPreprocessor': {
      options: {
        sourceMap: 'inline',
        presets: [ 'es2015-loose', 'stage-1'],
        plugins: [
          'syntax-flow',
          'transform-decorators-legacy',
          'transform-flow-strip-types'
        ]
      }
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false
  });
};
